src/java.base/share/classes/java/lang/runtime/StringTemplateImplFactory.java

```diff
@@ -0,0 +1,197 @@
+import java.util.List;
+
+import jdk.internal.javac.PreviewFeature;
+
+/**
+ * This class synthesizes {@link StringTemplate StringTemplates} based on
+ * fragments and bootstrap method type. Usage is primarily from
+ * {@link java.lang.runtime.TemplateRuntime}.
+ *
+ * @since 21
+ */
+@PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
+final class StringTemplateImplFactory {
+
+    /**
+     * Private constructor.
+     */
+    StringTemplateImplFactory() {
+        throw new AssertionError("private constructor");
+    }
+
+    /*
+     * {@link StringTemplateImpl} constructor MethodHandle.
+     */
+    private static final MethodHandle CONSTRUCTOR;
+
+    /**
+     * List (for nullable) of MethodHandle;
+     */
+    private static final MethodHandle TO_LIST;
+
+    static {
+        try {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+
+            MethodType mt = MethodType.methodType(void.class, int.class, int.class, List.class,
+                    MethodHandle.class, MethodHandle.class);
+            CONSTRUCTOR = lookup.findConstructor(StringTemplateImpl.class, mt)
+                    .asType(mt.changeReturnType(Carriers.CarrierObject.class));
+
+            mt = MethodType.methodType(List.class, Object[].class);
+            TO_LIST = lookup.findStatic(StringTemplateImplFactory.class, "toList", mt);
+        } catch(ReflectiveOperationException ex) {
+            throw new AssertionError("carrier static init fail", ex);
+        }
+    }
+
+    /**
+     * Create a new {@link StringTemplateImpl} constructor.
+     *
+     * @param fragments  string template fragments
+     * @param type       values types with StringTemplate return
+     *
+     * @return {@link MethodHandle} that can construct a {@link StringTemplateImpl} with arguments
+     * used as values.
+     */
+    static MethodHandle createStringTemplateImplMH(List<String> fragments, MethodType type) {
+        Carriers.CarrierElements elements = Carriers.CarrierFactory.of(type);
+        MethodHandle[] components = elements
+                .components()
+                .stream()
+                .map(c -> c.asType(c.type().changeParameterType(0, StringTemplateImpl.class)))
+                .toArray(MethodHandle[]::new);
+        Class<?>[] ptypes = elements
+                .components()
+                .stream()
+                .map(c -> c.type().returnType())
+                .toArray(Class<?>[]::new);
+        int[] permute = new int[ptypes.length];
+
+        MethodHandle interpolateMH;
+        MethodType mt;
+        try {
+            interpolateMH = StringConcatFactory.makeConcatWithTemplate(fragments, List.of(ptypes));
+        } catch (StringConcatException ex) {
+            throw new RuntimeException("constructing internal string template", ex);
+        }
+        interpolateMH = MethodHandles.filterArguments(interpolateMH, 0, components);
+        mt = MethodType.methodType(String.class, StringTemplateImpl.class);
```

Comment:
This MethodType can be stored in a static final field than created every time on the fly. Don't know if JIT compiler can inline this statement. Same fore that `List.class, StringTemplateImpl.class` type below.

Reply:
Changing